### PR TITLE
Added testExplorer and notebooks UI automated tests

### PR DESCRIPTION
**Pull request by srinathnarayanan**
This PR does the following
- Exposes a testExplorer landing page that hosts the explorer.html iframe
- Adds AAD authentication to get auth token to access notebooks and to use arm APIs to get the database account info
- This Info is used to initializae the data explorer iframe, similar to how the portal initializes it
- Adds a test that uploads, opens and deletes a notebook

### CODE REVIEW

File: src/Explorer/Explorer.ts [hunk 1]
```diff
@@ -1742,6 +1744,10 @@ export default class Explorer {
       return;
     }
 
+    if (event.data.data.type === MessageTypes.HideConnectScreen) {
```
Comment (southpolesteve) [inline src/Explorer/Explorer.ts:1747]: Why is this needed? If you start explorer with "platform=Portal" in the query string shouldn't it work?

File: src/TestExplorer.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
+import * as ko from "knockout";
```
Comment (southpolesteve) [inline src/TestExplorer.ts:1]: Can you rework this file without knockout? Would strongly prefer to avoid adding new knockout code.

File: src/TestExplorer.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      this.initTestExplorer();
+    };
+    window.addEventListener("message", this.handleMessage.bind(this), false);
+  }
+
+  private parseUrlParams = (): void => {
+    window.location.search
+      .substr(1)
+      .split("&")
+      .forEach(item => {
```
Comment (southpolesteve) [inline src/TestExplorer.ts:44]: This API is built in: new URLSearchParams(window.location.search)

File: src/TestExplorer.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        const tmp = item.split("=");
+        const value = decodeURIComponent(tmp[1]);
+        switch (tmp[0]) {
```
Comment (southpolesteve) [inline src/TestExplorer.ts:47]: If you use the above built in APIs, this should become much simpler. You can call `get` for each key

File: src/TestExplorer.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    };
+    window.postMessage(content, window.location.href);
+
+    const hideConnectContent = {
+      type: MessageTypes.HideConnectScreen
+    };
+    window.postMessage(hideConnectContent, window.location.href);
+  }
+
+  private sendMessageToExplorerFrame(data: unknown): void {
```
Comment (southpolesteve) [inline src/TestExplorer.ts:137]: Same as other file. Unclear why this is needed if loading in Portal mode.

File: src/testExplorer.html [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="height=device-height, width=device-width, initial-scale=1.0" />
+    <title>Azure Cosmos DB</title>
+    <link rel="shortcut icon" href="images/CosmosDB_rgb_ui_lighttheme.ico" type="image/x-icon" />
+  </head>
+
+  <body>
+    <switch-directory-pane params="{data: switchDirectoryPane}"></switch-directory-pane>
```
Comment (southpolesteve) [inline src/testExplorer.html:10]: Is this component needed? Again, trying to avoid new KO usage.

File: src/testExplorer.html [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+    <iframe id="explorerMenu" name="explorer" class="iframe" title="explorer" src="explorer.html?v=1.0.1&platform=Test">
```
Comment (southpolesteve) [inline src/testExplorer.html:12]: I see now you are using `platform=Test`. Is there a reason for needing a new one? I think we should be using platform=Portal so the underlying Explorer code knows what to do. We have a lot of checks for == Portal for behavior

File: test/notebooks/notebookTestUtils.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  const notebooksTestRunnerApplicationId = process.env.NOTEBOOKS_TEST_RUNNER_TENANT_ID;
+  const notebooksTestRunnerClientId = process.env.NOTEBOOKS_TEST_RUNNER_CLIENT_ID;
+  const notebooksTestRunnerClientSecret = process.env.NOTEBOOKS_TEST_RUNNER_CLIENT_SECRET;
+  const notebooksAccountName = process.env.NOTEBOOKS_ACCOUNT_NAME;
+  const notebooksAccountKey = process.env.NOTEBOOKS_ACCOUNT_KEY;
+  const notebooksAccountSubscriptonId = process.env.NOTEBOOKS_ACCOUNT_SUBSCRIPTION_ID;
+  const notebooksAccountResourceGroup = process.env.NOTEBOOKS_ACCOUNT_RESOURCE_GROUP;
+
+  const prodUrl = `https://localhost:1234/testExplorer.html?
+${TestExplorerParams.notebooksTestRunnerApplicationId}=${encodeURI(notebooksTestRunnerApplicationId)}&
```
Comment (southpolesteve) [inline test/notebooks/notebookTestUtils.ts:22]: You can do all of this with `URLSearchParams`:

```
const url = new URL("testExplorer.html", "https://localhost:1234");
url.searchParams.append(TestExplorerParams.notebooksTestRunnerApplicationId, notebooksTestRunnerApplicationId);
```

File: src/TestExplorer.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+      this.notebooksTestRunnerApplicationId,
+      this.notebooksTestRunnerClientId,
+      this.notebooksTestRunnerClientSecret
+    );
+    const token = await credentials.getToken("https://management.core.windows.net/.default");
+    return token.token;
+  }
+
+  private async getDatabaseAccount(token: string): Promise<DatabaseAccountsGetResponse> {
+    const client = new CosmosDBManagementClient(new CustomSigner(token), this.notebooksAccountSubscriptonId);
```
Comment (southpolesteve) [inline src/TestExplorer.ts:90]: Ugh. I can't believe our arm libraries don't work with the new @azure/identity library. Not your issue, but I will ping someone in central SDK about this. Frustrating.

File: test/notebooks/notebookTestUtils.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+export const NOTEBOOK_OPERATION_DELAY = 5000;
+export const RENDER_DELAY = 1000;
+
+let testExplorerFrame: Frame;
+export const getTestExplorerFrame = async (): Promise<Frame> => {
+  if (testExplorerFrame) {
+    return testExplorerFrame;
+  }
+
+  const notebooksTestRunnerApplicationId = process.env.NOTEBOOKS_TEST_RUNNER_TENANT_ID;
```
Comment (southpolesteve) [inline test/notebooks/notebookTestUtils.ts:13]: Where are all these `process.env` variables expected to be loaded from? Do they need to be added to .env.example so they can be run locally?

File: test/notebooks/notebookTestUtils.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
+import { ElementHandle, Frame } from "puppeteer";
```
Comment (southpolesteve) [inline test/notebooks/notebookTestUtils.ts:1]: Remove this file and inline these all in the test file. I think there is no benefit to abstracting this, especially for end to end tests. It adds another layer of indirection. Sandi Metz articulates my opinions on this very well https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction

Comment (southpolesteve) [reply test/notebooks/notebookTestUtils.ts:1]: By inlne, I mean make the test file one big function. No need to have these all broken up into small functions.

File: test/notebooks/notebookTestUtils.ts (hunk 10 above)
Comment (srinathnarayanan) [reply test/notebooks/notebookTestUtils.ts:13]: Yes, they need to be injected in the environment. WIll update .env.example file

File: src/Explorer/Explorer.ts (hunk 1 above)
Comment (srinathnarayanan) [reply src/Explorer/Explorer.ts:1747]: changed to "platform=Portal" and removed this.

File: src/TestExplorer.ts (hunk 5 above)
Comment (srinathnarayanan) [reply src/TestExplorer.ts:137]: changed to "platform=Portal" and removed this.

File: test/notebooks/notebookTestUtils.ts (hunk 11 above)
Comment (srinathnarayanan) [reply test/notebooks/notebookTestUtils.ts:1]: My thought was, for any notebook UI test, it should be framed as 
upload notebook -> do something (open / run / change kernel and run etc) -> delete notebook 
Hence this abstraction. Upload notebook and delete notebook should be called in the setup and cleanup for a class of tests. I will add one more test to this PR to make that clear.

File: test/notebooks/testExplorer/TestExplorer.ts [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+};
+
+const getDatabaseAccount = async (
+  token: string,
+  notebooksAccountSubscriptonId: string,
+  notebooksAccountResourceGroup: string,
+  notebooksAccountName: string
+): Promise<DatabaseAccountsGetResponse> => {
+  const client = new CosmosDBManagementClient(new CustomSigner(token), notebooksAccountSubscriptonId);
+  return await client.databaseAccounts.get(notebooksAccountResourceGroup, notebooksAccountName);
```
Comment (southpolesteve) [inline test/notebooks/testExplorer/TestExplorer.ts:49]:
```suggestion
  return client.databaseAccounts.get(notebooksAccountResourceGroup, notebooksAccountName);
```
No await needed. I should probably add a lint rule for this.

**Review by southpolesteve — APPROVED** [review]
Looking good! Thanks for tackling the test explorer setup. I plan to use that elsewhere.

File: test/notebooks/testNotebooks/GettingStarted.ipynb [hunk 13]
```diff
@@ -0,0 +1,110 @@
+{
+  "cells": [
```
Comment (swviswan) [inline test/notebooks/testNotebooks/GettingStarted.ipynb:2]: >"cells": [ [](start = 2, length = 10)

Nice. Can you rename this file to something so that if we see logs for this file we can quickly identify that it is our test and not the production NB.

**Review by swviswan — APPROVED** [review]
:shipit:

File: test/notebooks/testNotebooks/GettingStarted.ipynb (hunk 13 above)
Comment (srinathnarayanan) [reply test/notebooks/testNotebooks/GettingStarted.ipynb:2]: We wont see any logs here for running of the file. The errors/outputs will be saved to the file itself.
Running the file and parsing out outputs/checking for errors will be more elaborate, so pushed it to the next set of tests we may have to write.

Comment (srinathnarayanan) [reply test/notebooks/testNotebooks/GettingStarted.ipynb:2]: We will have to run the file, download it, parse the json and then validate results.
